core/src/main/kotlin/com/strumenta/kolasu/transformation/Transformation.kt

```diff
@@ -306,17 +374,86 @@ open class ASTTransformer(
 
+    inline fun <reified S : Any, reified T : Node> registerNodeFactory(): NodeFactory<S, T> {
+        return registerNodeFactory(S::class, T::class)
+    }
+
     fun <S : Any, T : Node> registerNodeFactory(source: KClass<S>, target: KClass<T>): NodeFactory<S, T> {
         registerKnownClass(target)
-        val nodeFactory = NodeFactory<S, T>({ _, _, _ ->
+        // We are looking for any constructor with does not take parameters or have default
+        // values for all its parameters
+        val emptyConstructor = target.constructors.find { it.parameters.all { param -> param.isOptional } }
+        val nodeFactory = NodeFactory({ source: S, _, f ->
             if (target.isSealed) {
                 throw IllegalStateException("Unable to instantiate sealed class $target")
             }
-            target.createInstance()
-        })
+            fun getConstructorParameterValue(kParameter: KParameter): ParameterValue {
+                try {
+                    val childNodeFactory = f.getChildNodeFactory<Any, T, Any>(target, kParameter.name!!)
+                    if (childNodeFactory == null) {
+                        if (kParameter.isOptional) {
+                            return AbsentParameterValue
+                        }
+                        throw java.lang.IllegalStateException(
+                            "We do not know how to produce " +
+                                "parameter ${kParameter.name!!} for $target"
+                        )
+                    } else {
+                        return when (val childSource = childNodeFactory.get.invoke(source)) {
+                            null -> {
+                                AbsentParameterValue
+                            }
+                            is List<*> -> {
+                                PresentParameterValue(childSource.map { transform(it) }.toMutableList())
+                            }
+
+                            is String -> {
+                                PresentParameterValue(childSource)
+                            }
+
+                            else -> {
+                                if (kParameter.type == String::class.createType() && childSource is ParseTree) {
+                                    PresentParameterValue(childSource.text)
+                                } else {
+                                    PresentParameterValue(transform(childSource))
+                                }
+                            }
+                        }
+                    }
+                } catch (t: Throwable) {
+                    throw RuntimeException(
+                        "Issue while populating parameter ${kParameter.name} in " +
+                            "constructor ${target.qualifiedName}.${target.preferredConstructor()}",
+                        t
+                    )
+                }
+            }
+            if (emptyConstructor == null) {
```

Comment:
First of all, thank you for your review. As usual, you have a great eye and this is very valuable.

If I understand correctly, we should check `childrenSetAtConstruction` because, while we set this value initially based on emptyConstructor being equal to null, this can be later changed in `withChild`, so we should really check the value that `childrenSetAtConstruction` time has when we actually invoke the factory. Is that right? I am pushing a change for this with a comment

For the second comment, I think I understand I should be doing this:
```
                if (emptyLikeConstructor == null) {
                    throw RuntimeException("childrenSetAtConstruction is set but there is no empty like " +
                            "constructor for $target")
                }
                target.createInstance()
```

I also added a few more comments to the code.

Thank you!